server/bleep/src/semantic/execute.rs

```diff
@@ -0,0 +1,68 @@
+use std::collections::HashMap;
+
+use crate::{
+    query::{
+        execute::{ApiQuery, PagingMetadata, QueryResponse, QueryResult, ResultStats},
+        parser::NLQuery,
+    },
+    snippet::Snippet,
+};
+
+use super::Semantic;
+
+use anyhow::Result;
+
+pub async fn execute(
+    semantic: Semantic,
+    query: NLQuery<'_>,
+    params: ApiQuery,
+) -> Result<QueryResponse> {
+    let data = semantic
+        .search(
+            &query,
+            params.page_size as u64,
+            ((params.page + 1) * params.page_size) as u64,
+        )
+        .await
+        .map(|raw| {
+            raw.into_iter()
+                .map(|v| super::Payload::from_qdrant(v.payload))
+                .collect::<Vec<_>>()
+        })?
+        .into_iter()
+        .fold(HashMap::new(), |mut acc, payload| {
+            acc.entry((
+                payload.relative_path.to_string(),
+                payload.repo_name.to_string(),
+                payload.repo_ref.to_string(),
+                Some(payload.lang.to_string()),
+            ))
+            .or_insert_with(Vec::new)
+            .push(Snippet {
+                data: payload.snippet.to_string(),
+                line_range: payload.start_line as usize..payload.end_line as usize,
+                highlights: vec![],
+                symbols: vec![],
+            });
+
+            acc
+        })
+        .into_iter()
+        .map(|((relative_path, repo_name, repo_ref, lang), snippets)| {
+            QueryResult::Snippets(crate::snippet::SnippedFile {
+                relative_path,
+                repo_name,
+                repo_ref,
+                snippets,
+                lang,
+            })
+        })
+        .collect::<Vec<_>>();
```

Comment:
We should add MMR deduplication. This will probably break the paging strategy, but we can implement the same approach that we take for Tantivy searches.